### DISCUSSION TITLE
ESP32 Cam Installation Howto - Gold Standard?

### DISCUSSION THREAD

**Discussion by belveder79** [arendst/Tasmota#18971] [Q&A]
So I'm working with ESP32 Cams for quite a while now, probably since the time Tasmota first supported the OV2640 and the webcam binaries. What I have found is that it becomes increasingly difficult to flash those devices and make them work reliable, and overall I feel that with the further development of Tasmota the situation got and gets worse and worse.

It does not matter which one you buy, they are all the same (imho):

[AIThinker](http://www.ai-thinker.com/pro_view-24.html)
[DIYMore](https://www.diymore.cc/products/esp32-cam-wifi-wireless-module-esp32-serial-to-wifi-esp32-cam-spi-flash-bluetooth-development-board-with-ov2640-camera-module)

I just recently bought 8 of the latter, and while the stock firmware works flawlessly (and I really mean FLAWLESSLY) and at high frame rates at any resolution, 6 out of 8 of those devices did not work at all (means give me any image any time) with any procedure I tried flashing Tasmota. There are lot of posts around 0x105 or 0xffffffff errors and usually people are referred to "get a better power supply". However, that is a prerequisite, yes, but that seems usually not to be the most prominent reason the cams don't work. And yes, the ESP32 quality sucks sometimes and you might find one in a few that is indeed dead. but it seems that there is either some step missing that I did not find yet, or it is an error related to me being a BDU.

The procedure is always the same:
 - unbox, erase mem with esptool
 - flash webcam-factory with esptool
 - boot
Flashing with the web installer does not work out in 90% of the cases (boot loop).

 I did some testing yesterday and I actually got 2 out of the 6 to work a bit, but far from satisfiable, and only with 9.5.0. Whether you use one or the other template for the ESP-Cam does not make any difference. If you get one to work (and I mean REALLY, it works rock-stable). However, I haven't found a plausible way to get to that point. 

Can someone share the gold standard way to flash those devices such that they work reliably? Thx...

**Comment by Jason2866** [comment]
With which device do you use to flash and how is the CAM powered during flash process?

**Comment by belveder79** [comment]
just plugging it in.... usually I use a Macbook Pro M1, also tried with Macbook Pro 2015, straight from the USB port, or via a powered USB3 hub... all with similar results... I can try using a separate USB2TTL converter if I disassemble it and you feel it makes sense...

**Comment by Jason2866** [comment]
Espcam boards do not have a onboard USB serial converter chip. So i asked.

**Comment by belveder79** [comment]
those boards all come with the additional board to power and flash them... but probably I can give it a try using extra power from a power plug?

**Reply by sfromis** [reply]
Not all ESP32-CAM boards come with that "MB" adapter board. My experience is that it is *horrible* hardware quality, and I only made flashing work by instead connecting another USB/serial adapter directly to the cam board.

**Comment by Jason2866** [comment]
Do you run the cam with the flash board still connected? Don't do this. GPIO 0 is used for CAM clock.
Nothing can be connected when running the cam sketch!

**Reply by belveder79** [reply]
hmmm... will give it a try!

**Comment by belveder79** [comment]
Ok, so I tried with the first of about 5 cams and your suggestion worked immediately. While flashing with this support board works well, powering it is not reliable. It is not only related to the GPIO0, it is more related to the overall voltage delivered, which seems to drop to below 4.5 even once tasmota starts up - directly connecting a USB cable to the 5V/Gnd works flawlessly... will continue investigating...

**Reply by belveder79** [reply]
So the most "reliable" way I found so far was soldering the red/black combination from an USB cable stripped from the original connector straight to the VCC/GND of the ESPCam - worked flawlessly...

**Comment by btsimonh** [comment]
It may not all be down to PSU.  Try the open PR (https://github.com/arendst/Tasmota/pull/19280).  I moved all the functionality inside a separate thread, and identified 0x105 as a time related non-reset issue on some boards where we don't have a reset line, only a power off line.
The exact issue is the ESP32 driver toggles the poweroff line for a very short off pulse to reset.  So I now explicitly trigger a longer off pulse if we get 0x105, and it starts every time (AI Thinker board).
There were also a a couple of possible conflicts (with flash/sd file write) which were partially protected, but that may also be improved.
If you do try it, please report in that PR?

**Reply by belveder79** [reply]
thx a lot... will do so asap...

**Reply by btsimonh** [reply]
I recently updated the branch to work with esp32-S3, fixed a bug relating to boards with no cam_pwdn line.
But the main thing that I found for my S3 board was that `wcclock 20` (the default) killed wifi performance.
If your cams work, but not so well, try `wcclock 26`, or `wcclock 13`.  This made a huge improvment of wifi performance on the Freenove FNK0085 board.  But changing the clock rate on my AIThinker style boards had little or no effect.

**Reply by btsimonh** [reply]
p.s. If you like the new driver, please comment in the PR.  And if you are looking to do timelapse or motion detection, let's talk?

**Comment by belveder79** [comment]
I have installed the latest development snapshot and the new driver works really decently... thx a lot...
